### ISSUE TITLE
Support Python 3?

### ISSUE THREAD

**Issue by saschanaz** [tabatkins/bikeshed#1138]
Python 2 support [will end in 2020](https://www.python.org/dev/peps/pep-0373/#maintenance-releases):

>Being the last of the 2.x series, 2.7 will have an extended period of maintenance. The current plan is to support it for at least 10 years from the initial 2.7 release. This means there will be bugfix releases until 2020.

bikeshed probably wants to support Python 3 before the end of Py2 (unless there comes another extended life), thoughts?

**Comment by tabatkins** [comment]
Yup, I plan to move to Python 3 at some point. Right now I'm mostly held back by the Shepherd system, which uses Bikeshed, and has a significant Python2 set of related tools.

**Comment by cynthia** [comment]
There is a partial patch that makes it at least run under Py3 here: #1133 

Apparently Travis does not approve of this, not sure if I can commit to spend more time on it unless this can wait.

**Comment by saschanaz** [comment]
>Right now I'm mostly held back by the Shepherd system

Is the Shepherd system open sourced so that anyone can submit a Py3 patch?

**Comment by cynthia** [comment]
@plinss can probably answer that question.

**Comment by cynthia** [comment]
If a PR is made to allow Bikeshed to support both Python 2 and 3 (and CI ensures that compatibility doesn’t break) - as I understand the problem wouldn’t break Shepherd. Getting the plugin system to work on Py3 was more work than I could commit to - but feel free to take the PR from where it is right now.

**Comment by plinss** [comment]
Shepherd uses internal mercurial apis, so it's restricted to Python2 until mercurial and needed extensions, like hg-git, are Python3 friendly (which was in progress last time I looked, but I don't have the bandwidth these days to test thoroughly).

That said, the only cross dependencies between Bikeshed and Shepherd are [widlparser](https://github.com/plinss/widlparser) and [apiclient](https://github.com/plinss/apiclient). Both are open source and patches to make those work on both Python2 and Python3 would be welcome. Theoretically this should be straightforward with six, but last time I tried I was having some issues with the string objects in widlparser. It might require some reworking of the widlparser apis.

(The rest of Shepherd is open source as well, but not currently on GitHub.)

**Comment by sideshowbarker** [comment]
It seems worth noting here that Python 2 will be “[completely dead on Jan 1 2020](https://mail.python.org/pipermail/python-dev/2018-March/152355.html)”.

So we have about 9 months left until then — until Python 2.7 stops being maintained;  specifically, until Python 2.7. officially stops gets any bug fixes — “[not even source-only security patches](https://mail.python.org/pipermail/python-dev/2018-March/152348.html)“.

https://www.python.org/dev/peps/pep-0373/#maintenance-releases
> Being the last of the 2.x series, 2.7 will have an extended period of maintenance. Specifically, 2.7 will receive bugfix support until January 1, 2020. After the last release, 2.7 will receive no support.

So it might be worth considering to escalate/(re)prioritize this issue with that end-of-life date in mind  — or maybe even add a “help wanted” label or whatever, to get additional time into from other contributors if needed.

**Comment by dtig** [comment]
Python 2.7 end-of-life is next month, checking in to see if Bikeshed will move to Python 3 soon?

**Comment by tabatkins** [comment]
It will! Probably in the early part of next year, as I'll soon be heading to vacation until the new year. Python 2 won't immediately explode on Jan 1st (nor will `pip` stop supporting it immediately), so we should be fine with a small delay.

**Comment by gunta** [comment]
From brew:
```
No available formula with the name "python@2"

python@2 was deleted from homebrew/core in commit 028f11f9:
  python@2: delete (https://github.com/Homebrew/homebrew-core/issues/49796)
  EOL 1 January 2020.
  We gave it 1 month more to live so that people had time to migrate.
  All in all, developers had 11 years to do their migration.
```

**Comment by tabatkins** [comment]
Ah, phew, ok I'll finish the branch. ^_^

**Comment by WimLeflere** [comment]
Can we help?

**Comment by tabatkins** [comment]
All right, I've got it working in a pyenv-installed 3.7 environment. Still a few fixes needed to get it running the full testsuite, because one of my dependencies changed from camelCase methods to under_case, but otherwise it's more or less done.

So fingers crossed that it'll be ready for merging into master by Friday.

**Comment by plinss** [comment]
Let me know when it's ready to merge into master, I'll need to update the csswg server to have a 3.7 venv to run bikeshed in before it pulls the new version. Or you can disable the webhook until I update the server.

It'll be easier for me to update the server on the weekend.

**Comment by tabatkins** [comment]
The Python 3 branch was merged into master last night, after I finally greened all the tests.

Install instructions forthcoming, but the important bit is that it requires Python 3.7, which is not the version installed by default on many OSes. I recommend using pyenv to install it, as it's very easy to use. [Here's a pyenv tutorial.](https://amaral.northwestern.edu/resources/guides/pyenv-tutorial)

Once that's in, a `pip install --editable .` in the repo should work as it normally does.
